cmd/kubelet/app/server.go

```diff
@@ -513,6 +537,47 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 	return nil
 }
 
+func initializeCertificateManager(certDirectory string, certRotation int32, nodeName types.NodeName, certData []byte, keyData []byte) (certificate.Manager, error) {
+	certificateStore, err := certificate.NewFileStore(
+		"kubelet-client",
+		certDirectory,
+		certDirectory,
+		"",
+		"")
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize certificate store: %v", err)
+	}
+	var rotation uint
+	if certRotation > 0 {
+		rotation = uint(certRotation)
+	}
+	clientCertificateManager, err := certificate.NewManager(&certificate.Config{
+		Template: &x509.CertificateRequest{
+			Subject: pkix.Name{
+				Organization: []string{"system:nodes"},
+				CommonName:   fmt.Sprintf("system:node:%s", nodeName),
+			},
+		},
+		Usages: []certificates.KeyUsage{
+			certificates.UsageDigitalSignature,
+			certificates.UsageKeyEncipherment,
+			certificates.UsageClientAuth,
+		},
+		CertificateStore:           certificateStore,
+		CertificateRotationPercent: rotation,
+		Rotation: func() {
+			glog.Info("The client certificate was rotated, restarting so changes take effect.")
+			os.Exit(0)
```

Comment:
I agreed with @timstclair on this. Restarting Kubelet is very disruptive. In addition to the issues pointed out by @timstclair, there are other concerns:

1) In GKE, we might agree upon O(months) certification rotation, but how we ensure this behavior for OSS users? 
2) Any misbehaviour operation on certification rotation might lead to a node readiness flapping. But from the users' viewpoint, they only see node status is changed between ready and notReady unless they view the log. This adds more debuggability and introspection difficulties. 
3) We encountered a lot of issues with mystery kubelet / docker restarts due to various reasons in prod. This feature only makes things worse. 
4) What about other standalone daemons running on the node, which are talking to API server too, for example, NPD in GKE. Based on this design, are we going to force NPD restart too?